components/sup/src/manager/file_watcher.rs

```diff
@@ -1,3464 +1,2652 @@
+        let watcher = SupWatcher::new(tx, config).map_err(Error::NotifyCreateError)?;
+        let watched_file = WatchedFile::new(&path.into())?;
+        let send_initial_event = !ignore_initial;
+        let mut file_watcher = Self { callbacks,
+                                      watcher,
+                                      rx,
+                                      watched_file,
+                                      send_initial_event };
+        file_watcher.watch()?;
+        Ok(file_watcher)
+    }
+
+    fn manage_watch(&mut self) -> Result<()> {
+        self.manage_watch_with_next_state_as(self.watched_file.assess_state())
+    }
+
+    fn manage_watch_with_next_state_as(&mut self, next_state: WatchedFileState) -> Result<()> {
+        trace!("FileWatcher::manage_watch_with()");
+        self.unwatch()?;
+        self.watched_file.state = next_state;
+        self.watch()
+    }
+
+    /// The point of interface between our system and the notify crate where we register watches
+    /// based on our most recent observation of our WatchedFile.
+    fn watch(&mut self) -> Result<()> {
+        trace!("FileWatcher::watch()");
+        match &self.watched_file.state {
+            WatchedFileState::ExistentFile => {
+                trace!("| WatchedFileState::ExistentFile");
+                self.watcher
+                    .watch(&self.watched_file.path.clone(), RecursiveMode::NonRecursive)
+                    .map_err(Error::NotifyError)?;
+            }
+            WatchedFileState::ExistentDirectory => {
+                trace!("| WatchedFileState::ExistentDirectory");
+                self.watcher
+                    .watch(&self.watched_file.directory.clone(),
+                           RecursiveMode::NonRecursive)
+                    .map_err(Error::NotifyError)?;
+            }
+            WatchedFileState::NonExistent => {
+                trace!("| WatchedFileState::NonExistent");
+            }
+        }
+        Ok(())
     }
 
-    fn get_chain_link_info(&self) -> ChainLinkInfo {
-        ChainLinkInfo { path: self.path.clone(),
-                        prev: self.prev.clone(), }
+    /// The point of interface between our system and the notify crate where we clear watches
+    /// based on our most recent observation of our WatchedFile.
+    fn unwatch(&mut self) -> Result<()> {
+        trace!("FileWatcher::unwatch()");
+        match &self.watched_file.state {
+            WatchedFileState::ExistentFile => {
+                trace!("| WatchedFileState::ExistentFile");
+                match self.watcher.unwatch(&self.watched_file.path.clone()) {
+                    Ok(()) => (),
+                    Err(e) => {
+                        match e.kind {
+                            notify::ErrorKind::WatchNotFound => (),
+                            _ => return Err(Error::NotifyError(e)),
+                        }
+                    }
+                }
```

Comment:
this could be simplified to:
```
                if let Err(e) = self.watcher.unwatch(&self.watched_file.path.clone()) {
                    if !matches!(e.kind, notify::ErrorKind::WatchNotFound) {
                        return Err(Error::NotifyError(e));
                    }
                }
```

Reply:
Done.